main/guides/agoric-cli/agd-query-tx.md

```diff
@@ -0,0 +1,167 @@
+# Using `agd` to make queries and transactions
+
+`agd` is the Agoric Cosmos App, analagous to `simd` in the [Cosmos SDK](https://docs.cosmos.network/) simapp or `gaiad` in
+the Cosmos hub. Most of the `simd` [query commands](https://docs.cosmos.network/v0.46/core/cli.html#query-commands) and [transaction commands](https://docs.cosmos.network/v0.46/core/cli.html#transaction-commands) work similarly in `agd`.
+
+::: tip agd for Building Dapps
+
+This section focusses on commands relevant to developing and deploying smart contracts and dapps. See also:
+
+- [Validators topics \- Agoric Community Forum](https://community.agoric.com/c/validators/9)
+- [Governance topics \- Agoric Community Forum](https://community.agoric.com/c/governance/6)
+- [Delegator Guide \(CLI\) \| Cosmos Hub](https://hub.cosmos.network/main/delegators/delegator-guide-cli.html)
+
+:::
+
+::: tip Installing agd
+
+Options include:
+
+- Use the [basic dapp local chain](../getting-started/#starting-a-local-agoric-blockchain) docker container: to run `agd status`, enter `yarn docker:bash` followed by `agd status`; or use `docker-compose exec agd agd status`.
+- Install from an [agoric-sdk release](https://github.com/Agoric/agoric-sdk/releases).
+
+:::
+
+If we invoke `agd` without arguments, it prints a list of available commands including:
+
+```
+Available Commands:
+
+  help                Help about any command
+  keys                Manage your application's keys
+  query               Querying subcommands
+  status              Query remote node for status
+  tx                  Transactions subcommands
+  version             Print the application binary version information
+
+Flags:
+  -h, --help                help for agd
+      --home string         directory for config and data (default $HOME)
+```
+
+## Query Commands
+
+In most cases, `agd query ...` is followed by a module name such as `bank`. An exception is `agd status`:
+
+### agd status
+
+Query remote node for status
+
+Example:
+
+```console
+$ agd status
+{"NodeInfo":{"protocol_version":{"p2p":"8","block":"11" ... }}}
+```
+
+The query goes to a local node at `tcp://localhost:26657` by default. To use another node:
+
+```console
+$ agd status --node https://devnet.rpc.agoric.net:443
+{"NodeInfo":{"protocol_version":{"p2p":"8","block":"11" ... }}}
+```
+
+::: tip Port is required
+
+Typically, `:443` can be left implicit in `https` URLs.
+But not here. Without it, we get:
```

Comment:
I don't know why cosmos SDK requires it. I just know that it does.
@JimLarson ? @michaelfig ? Do you know? Should we open an issue?

Reply:
Handling for the `--node $rpcUri` option is [`client, err := NewClientFromNode(rpcURI)`](https://github.com/cosmos/cosmos-sdk/blob/77b2366819931c1c2478eee80db413ed7c2a5a77/client/cmd.go#L147), where [`NewClientFromNode(nodeURI string)` returns `rpchttp.New(nodeURI, "/websocket")`](https://github.com/cosmos/cosmos-sdk/blob/77b2366819931c1c2478eee80db413ed7c2a5a77/client/utils.go#L78-L82) and `rpchttp` is now [cometbft rpc/client/http/http.go](https://github.com/cometbft/cometbft/blob/f72d930a68386f6139838449d0653ef0621f7b29/rpc/client/http/http.go#L110-L118) but was formerly [tendermint rpc/client/http/http.go](https://github.com/tendermint/tendermint/blob/35581cf54ec436b8c37fabb43fdaa3f48339a170/rpc/client/http/http.go#L108-L117). Both use [url.Parse](https://pkg.go.dev/net/url#Parse) to parse the input string into a `parsedUrl` struct "u" and return an [http.Client](https://pkg.go.dev/net/http#Client) with a [Transport](https://pkg.go.dev/net/http#Transport) whose Dial function is implemented to return `net.Dial(protocol, u.GetDialAddress())` [_[CometBFT](https://github.com/cometbft/cometbft/blob/f72d930a68386f6139838449d0653ef0621f7b29/rpc/jsonrpc/client/http_json_client.go#L403-L422), [Tendermint](https://github.com/tendermint/tendermint/blob/35581cf54ec436b8c37fabb43fdaa3f48339a170/rpc/jsonrpc/client/http_json_client.go#L387-L406)_], where `GetDialAddress()` for a URL destination [_as opposed to the path of a Unix domain socket_] returns the parsed `Host` (causing failure in the absence of a port because [net.Dial](https://pkg.go.dev/net#Dial) is low-level and requires that).

Bottom line, this could easily be fixed by updating `func (u parsedURL) GetDialAddress()` to pay attention to URL Scheme, e.g.
```diff
 // GetDialAddress returns the endpoint to dial for the parsed URL.
 func (u parsedURL) GetDialAddress() string {
-	// if it's not a unix socket we return the host, example: localhost:443
+	// if it's not a unix socket we return the host with port, example: localhost:443
 	if !u.isUnixSocket {
+		hasPort, err := regexp.MatchString(`:[0-9]+$`, u.Host)
+		if err == nil && !hasPort {
+			if u.Scheme == protoHTTP || u.Scheme == protoWS {
+				return u.Host + `:80`
+			} else if u.Scheme == protoHTTPS || u.scheme == protoWSS {
+				return u.Host + `:443`
+			}
+		}
 		return u.Host
 	}
 	// otherwise we return the path of the unix socket, ex /tmp/socket
 	return u.GetHostWithPath()
 }
```
I think it makes sense to report upstream.

Reply:
https://github.com/cometbft/cometbft/issues/1902

Reply:
I'm guessing that since they're using the URL format, they don't want to mess with the assumption that 80 is the implicit port number. Since they don't want to squat on the webserver's favorite port, it would make sense to pick a different standard port number, then make it explicit in the node argument.

Reply:
Update: cometbft no longer requires an explicit port for http/https or ws/wss URLs.